core/chunk/src/lib.rs

```diff
@@ -244,31 +257,85 @@ impl Chunk {
         &self.heightmaps
     }
 
-    fn update_heightmap(&mut self, x: usize, y: usize, z: usize, block: BlockId) -> HeightMapMask {
-        let heightmap = self.heightmap_mut(x, z);
+    fn update_heightmap(
+        &mut self,
+        x: usize,
+        y: usize,
+        z: usize,
+        old_block: BlockId,
+        new_block: BlockId,
+    ) -> HeightMapMask {
         let mut mask: HeightMapMask = HeightMapMask::empty();
-        if (block.is_solid() || block.is_fluid()) && heightmap.motion_blocking() < y as u8 {
-            heightmap.set_motion_blocking(y as u8);
-            mask |= HeightMapMask::MOTION_BLOCKING;
-        }
+        let y = y as u16;
 
-        if (block.is_solid() || block.is_fluid())
-            && !block.is_leaves()
-            && heightmap.motion_blocking_no_leaves() < y as u8
-        {
-            heightmap.set_motion_blocking_no_leaves(y as u8);
-            mask |= HeightMapMask::MOTION_BLOCKING_NO_LEAVES;
-        }
+        struct HeightMapCheckContext(
+            &'static dyn Fn(BlockId) -> bool,
+            HeightMapMask,
+            &'static dyn Fn(&HeightMap) -> u16,
+            &'static dyn Fn(&mut HeightMap, u16) -> (),
+        );
 
-        if block.is_solid() && heightmap.ocean_floor() < y as u8 {
-            heightmap.set_ocean_floor(y as u8);
-            mask |= HeightMapMask::OCEAN_FLOOR;
-        }
+        let checks: Vec<HeightMapCheckContext> = vec![
```

Comment:
I'm not so sure about this allocation—this is potentially very hot code. It looks to me that using a fixed array instead of a `Vec` should work here.

Reply:
Well, I mean optimally this whole vec should be optimized out and the loop unrolled with the functions inlined, I just did it like this to avoid redundant code. I'll change it to an array but I doubt it'll make a difference after optimization.